micro-rdk/src/common/data_manager.rs

```diff
@@ -231,13 +247,68 @@ pub enum DataSyncError {
     DataStoreError(#[from] DataStoreError),
     #[error(transparent)]
     MessageDecodingError(#[from] prost::DecodeError),
+    #[error("time correction resulted in out of bounds duration")]
+    TimeOutOfBoundsError,
+    #[error("current time unset")]
+    NoCurrentTime,
 }
 
 pub struct DataSyncTask<StoreType> {
     store: Rc<AsyncMutex<StoreType>>,
     resource_method_keys: Vec<ResourceMethodKey>,
     sync_interval: Duration,
     part_id: String,
+    robot_start_time: Instant,
+}
+
+fn get_time_to_subtract(
+    robot_start_time: Instant,
+    stored_time: Timestamp,
+) -> Result<chrono::Duration, DataSyncError> {
+    let stored_time_dur = Duration::new(stored_time.seconds as u64, stored_time.nanos as u32);
+    let time_to_subtract = robot_start_time.elapsed() - stored_time_dur;
+    let time_to_subtract = chrono::Duration::new(
+        time_to_subtract.as_secs() as i64,
+        time_to_subtract.subsec_nanos(),
+    )
+    .ok_or(DataSyncError::TimeOutOfBoundsError)?;
+    Ok(time_to_subtract)
+}
+
+fn time_corrected_reading_handler(
+    robot_start_time: Instant,
+) -> impl FnMut(BytesMut) -> Result<SensorData, DataSyncError> {
```

Comment:
I'd expect the returned function to capture `robot_start_time` by value, so maybe it can just be `Fn` ?

Reply:
I'm not sure I understand what you mean?

Reply:
if it's to change the signature, I had to make it return `FnMut` so it could be provided to `map`

Reply:
OK, yeah I see that `std::vec::Drain::map` declares that it takes `FnMut`: https://doc.rust-lang.org/std/vec/struct.Drain.html#method.map, but shouldn't you always be able to pass a `Fn` where a `FnMut` is needed, since `FnMut` is a supertrait of `Fn`?

Reply:
but an `Fn` can't be guaranteed to satisfy `FnMut`, only vice versa, right?

Reply:
I'm getting myself confused now, or maybe I was originally. But this language from https://doc.rust-lang.org/std/ops/trait.Fn.html is what I was going by:

> Since both [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) and [FnOnce](https://doc.rust-lang.org/std/ops/trait.FnOnce.html) are supertraits of Fn, any instance of Fn can be used as a parameter where a [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) or [FnOnce](https://doc.rust-lang.org/std/ops/trait.FnOnce.html) is expected.

So if `Drain::map` wants a `FnMut` a `Fn` should be fine, and then this factory should be able to return `Fn`.

But really, this is unimportant. If the compiler is telling you it needs to be `FnMut`, then by all means listen to the compiler and ignore me.

Reply:
Oh I misunderstood how that works, `Fn` worked fine, changing